plugins/cairo-lang-quote/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+            (Delimiter::Parenthesis, false) => Self::Content("(".to_string()),
+            (Delimiter::Parenthesis, true) => Self::Content(")".to_string()),
+            (Delimiter::None, _) => Self::Content(String::default()),
+        }
+    }
+}
+
+fn process_token_stream(
+    mut token_stream: Peekable<impl Iterator<Item = RustTokenTree>>,
+    output: &mut Vec<QuoteToken>,
+) {
+    // Rust proc macro parser to TokenStream gets rid of all whitespaces.
+    // Here we just make sure no two identifiers are without a space between them.
+    let mut was_last_ident: bool = false;
+    while let Some(token_tree) = token_stream.next() {
+        match token_tree {
+            RustTokenTree::Group(group) => {
+                let token_iter = group.stream().into_iter().peekable();
+                let delimiter = group.delimiter();
+                output.push(QuoteToken::from_delimiter(delimiter, false));
+                process_token_stream(token_iter, output);
+                output.push(QuoteToken::from_delimiter(delimiter, true));
+                was_last_ident = false;
+            }
+            RustTokenTree::Punct(punct) => {
+                if punct.as_char() == '#' {
+                    if let Some(RustTokenTree::Ident(ident)) = token_stream.next() {
+                        let var_ident = Ident::new(&ident.to_string(), Span::call_site());
+                        output.push(QuoteToken::Var(var_ident))
+                    }
+                } else {
+                    output.push(QuoteToken::Content(punct.to_string()));
+                }
+                was_last_ident = false;
+            }
+            RustTokenTree::Ident(ident) => {
+                if was_last_ident {
+                    output.push(QuoteToken::Content(format!(" {}", ident)));
+                } else {
+                    output.push(QuoteToken::Content(ident.to_string()));
+                }
+                was_last_ident = true;
+            }
+            RustTokenTree::Literal(literal) => {
+                output.push(QuoteToken::Content(literal.to_string()));
+                was_last_ident = false;
+            }
+        }
+    }
+}
+
+#[proc_macro]
+pub fn quote(input: RustTokenStream) -> RustTokenStream {
+    let mut parsed_input: Vec<QuoteToken> = Vec::new();
+    let mut output_token_stream = rust_quote! {
+      use cairo_lang_macro::{TokenTree, Token, TokenStream};
+      use cairo_lang_primitive_token::ToPrimitiveTokenStream;
+      let mut quote_macro_result = TokenStream::empty();
+    };
+
+    let token_iter = input.into_iter().peekable();
+    process_token_stream(token_iter, &mut parsed_input);
+
+    for quote_token in parsed_input.iter() {
+        match quote_token {
+            QuoteToken::Content(content) => {
+                output_token_stream.extend(rust_quote! {
+                  quote_macro_result.push_token(TokenTree::Ident(Token::new(#content.to_string(), None)));
```

Comment:
Use full paths, as this can break user imports. Paths starting with `::` are recognized as absolute paths starting from some crate root rather than a module. Add tests for colliding `use` statements in user code.

```rs
use my_crate::my_module::Token;

fn my_macro(){
    quote!(some content)
}
```
where `Token::new` has different signature would fail to compile

Reply:
https://doc.rust-lang.org/reference/paths.html#path-qualifiers

The current code can still break if there is `mod cairo_lang_macro` in user code and `cairo_lang_macro` crate is used with `as` alias. Prefix it with `::` to make sure that rustc will interpret `::cairo_lang_macro` as crate and NOT module.